File: tree/leetcode144_preorder/recursion.js
// 144. Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/description/
// Given the root of a binary tree, return the preorder traversal of its nodes' values.

// Preorder: root -> left subtree -> right subtree

// Recursive
var preorderTraversal = function (root) {
  if (!root) return [];
  return [
    root.val,
    ...preorderTraversal(root.left),
    ...preorderTraversal(root.right),
  ];
};

File: tree/leetcode144_preorder/recursion.py
# 144. Binary Tree Preorder Traversal
# https://leetcode.com/problems/binary-tree-preorder-traversal/description/
# Given the root of a binary tree, return the preorder traversal of its nodes' values.

# Preorder: root -> left subtree -> right subtree


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution2(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        if root is None:
            return []

        return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right)
